src/StatsdClient/Worker/ConcurrentQueueWithPool.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace StatsdClient.Worker
+{
+    /// <summary>
+    /// ConcurrentQueueWithPool is a concurrent queue that also provides a pool of objects.
+    /// The expected worflow is the following:
+    /// Sender:
+    ///     - Get an object from the pool with TryDequeueFromPool
+    ///     - Initialize the object
+    ///     - Enqueue the object in the queue with EnqueueValue
+    /// Receiver:
+    ///     - Get an object from the queue with TryDequeueValue
+    ///     - Use the object
+    ///     - When the object is not used, put the object in the pool with EnqueuePool.
+    /// </summary>
+    internal class ConcurrentQueueWithPool<T>
+    where T : new()
```

Comment:
The `new()` generic constraint internally calls `Activator.CreateInstance`, which is slow (and allocates). You should instead use a `Func<T>` factory to eliminate that cost

Reply:
Fixed.